Source/WebCore/html/SearchInputType.cpp

```diff
@@ -146,7 +146,7 @@ auto SearchInputType::handleKeydownEvent(KeyboardEvent& event) -> ShouldCallBase
     const String& key = event.keyIdentifier();
     if (key == "U+001B"_s) {
         Ref<HTMLInputElement> protectedInputElement(*element());
-        protectedInputElement->setValueForUser(emptyString());
+        protectedInputElement->setValue({ }, DispatchChangeEvent);
```

Comment:
It seems strange to stop using `setValueForUser`, given that clearing the value here is a user-driven action. Also I think we need to fire an `input` event here as `input` events should fire whenever value changes occur.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event

Reply:
OK, I can switch it back and update the expectations of fast/forms/search-cancel-button-events.html instead.

Instead of:

```
Click the cancel button
PASS search.value is ""
PASS changeEventCounter is 0
PASS inputEventCounter is 1
```

We'll have:

```
Click the cancel button
PASS search.value is ""
PASS changeEventCounter is 1
PASS inputEventCounter is 1
```

0 looked better to me because the value there didn't really change. It's a little confusing because the code change looks like it's removing an input event dispatch, not a change event, and I don't know why that happens.

Reply:
Actually, I think this change is fine as-is.

Here is the logic in `TextFieldInputType`:

```
    case DispatchChangeEvent:
        // If the user is still editing this field, dispatch an input event rather than a change event.
        // The change event will be dispatched when editing finishes.
        if (input->focused())
            input->dispatchFormControlInputEvent();
        else
            input->dispatchFormControlChangeEvent();
        break;
```

There is a change in the test case, it the value changes from "foo" to "".

Pressing the cancel button focuses the element, so `DispatchChangeEvent` will fire an `input` event, but not a `change` event immediately. The `change` event will then get fired when the element is blurred.

Our current behavior also matches Chrome.

Reply:
Uh, I might need to investigate this more, because it seems the test is still failing even though what I have now should be equivalent to what was there before.

Reply:
`{ }` is a null string, whereas there was an empty string before – I think your latest change should work.

Reply:
Huh... that's a kind of important difference. I probably should have learned that sooner than now.

Reply:
It didn't work. O_O

Reply:
Which is weird, because it sure looks like these calls should now be completely identical. We must be missing something.

Reply:
Ahhh, I missed a spot! `SearchFieldCancelButtonElement::defaultEventHandler` in TextControlInnerElements.cpp. That must be it.